File: everything_parsing/11-nfa-epsilon/java/NfaTest.java
import java.util.List;
import java.util.Set;

/** java -cp java/out NfaTest */
public final class NfaTest {
    public static void main(String[] args) {
        System.out.println("NfaTest");

        // ── nondeterminism: two transitions on the same symbol ──
        Nfa abb = Nfa.parse("""
            states: S S1 S2 S3
            alphabet: a b
            start: S
            accept: S3
            S a S
            S b S
            S a S1
            S1 b S2
            S2 b S3
            """);
        Assert.equals(abb.move("S", "a").size(), 2, "S has two targets on 'a'");
        Assert.that(abb.acceptsString("abb"), "abb");
        Assert.that(abb.acceptsString("aabb"), "aabb");
        Assert.that(abb.acceptsString("bababb"), "bababb");
        Assert.that(!abb.acceptsString("ab"), "ab is not accepted");
        Assert.that(!abb.acceptsString("abba"), "abba is not accepted");
        Assert.that(!abb.acceptsString(""), "empty is not accepted");

        // trace: the state SET grows and shrinks
        var tr = abb.trace(List.of("a", "a", "b", "b"));
        Assert.equals(Nfa.showSet(tr.get(0)), "{S}", "start set");
        Assert.equals(Nfa.showSet(tr.get(1)), "{S,S1}", "after first a");
        Assert.equals(Nfa.showSet(tr.get(4)), "{S,S3}", "after aabb -- S3 reached");

        // ── epsilon closure ──
        Nfa eps = Nfa.parse("""
            states: I U A1 A2 B1 B2 W F
            alphabet: a b
            start: I
            accept: F
            I epsilon U
            I epsilon F
            U epsilon A1
            U epsilon B1
            A1 a A2
            B1 b B2
            A2 epsilon W
            B2 epsilon W
            W epsilon U
            W epsilon F
            """);
        Assert.that(eps.hasEpsilon(), "the NFA has epsilon transitions");
        Assert.equals(Nfa.showSet(eps.epsilonClosure(Set.of("I"))), "{A1,B1,F,I,U}",
            "epsilon-closure of the start includes F (so '' is accepted)");
        Assert.that(eps.acceptsString(""), "(a|b)* accepts the empty string");
        Assert.that(eps.acceptsString("ab") && eps.acceptsString("ba") && eps.acceptsString("abbaab"),
            "(a|b)* accepts mixed strings");

        // ── epsilon elimination preserves the language ──
        Nfa noEps = eps.removeEpsilon();
        Assert.that(!noEps.hasEpsilon(), "removeEpsilon() leaves no epsilon transitions");
        Assert.equals(eps.language(6), noEps.language(6),
            "the epsilon-free NFA accepts exactly the same language (up to length 6)");
        Assert.that(noEps.accept.contains("I"), "I is accepting now (its closure reached F)");

        // ── NFA for (a|b)*abb == DFA (Module 10) == regex ──
        Dfa dfa = Dfa.parse("""
            states: S A AB ABB
            alphabet: a b
            start: S
            accept: ABB
            S a A
            S b S
            A a A
            A b AB
            AB a A
            AB b ABB
            ABB a A
            ABB b S
            """);
        Regex re = Regex.parse("(a|b)*abb");
        for (String w : abb.language(7)) {
            if (w.equals("epsilon")) continue;
            Assert.that(dfa.acceptsString(w), "DFA agrees on \"" + w + "\"");
            Assert.that(re.matches(w), "regex agrees on \"" + w + "\"");
        }
        for (String w : dfa.language(7))
            if (!w.equals("epsilon"))
                Assert.that(abb.acceptsString(w), "NFA agrees on \"" + w + "\"");

        // ── malformed ──
        Assert.that(rejects("states: S\nstart: S\nS a S"), "missing alphabet:");
        Assert.that(rejects("states: S\nalphabet: a\nstart: S\nS a Q"), "transition to unknown state");

        Assert.summary();
    }

    private static boolean rejects(String s) {
        try { Nfa.parse(s); return false; } catch (IllegalArgumentException e) { return true; }
    }
}

File: everything_parsing/11-nfa-epsilon/js/nfa.test.mjs
// node js/nfa.test.mjs

import { that, equals, summary } from "./assert.mjs";
import { Nfa, showSet } from "./nfa.mjs";
import { Dfa } from "./dfa.mjs";
import { parse as parseRegex, matches } from "./regex.mjs";

console.log("nfa.test.mjs");

// ── nondeterminism ──
const abb = Nfa.parse(
  "states: S S1 S2 S3\nalphabet: a b\nstart: S\naccept: S3\n" +
  "S a S\nS b S\nS a S1\nS1 b S2\nS2 b S3\n");
equals(abb.move("S", "a").size, 2, "S has two targets on 'a'");
that(abb.acceptsString("abb"), "abb");
that(abb.acceptsString("aabb"), "aabb");
that(abb.acceptsString("bababb"), "bababb");
that(!abb.acceptsString("ab"), "ab is not accepted");
that(!abb.acceptsString("abba"), "abba is not accepted");
that(!abb.acceptsString(""), "empty is not accepted");

const tr = abb.trace([..."aabb"]);
equals(showSet(tr[0]), "{S}", "start set");
equals(showSet(tr[1]), "{S,S1}", "after first a");
equals(showSet(tr[4]), "{S,S3}", "after aabb -- S3 reached");

// ── epsilon closure ──
const eps = Nfa.parse(
  "states: I U A1 A2 B1 B2 W F\nalphabet: a b\nstart: I\naccept: F\n" +
  "I epsilon U\nI epsilon F\nU epsilon A1\nU epsilon B1\nA1 a A2\nB1 b B2\n" +
  "A2 epsilon W\nB2 epsilon W\nW epsilon U\nW epsilon F\n");
that(eps.hasEpsilon(), "the NFA has epsilon transitions");
equals(showSet(eps.epsilonClosure(new Set(["I"]))), "{A1,B1,F,I,U}",
  "epsilon-closure of the start includes F");
that(eps.acceptsString(""), "(a|b)* accepts the empty string");
that(eps.acceptsString("ab") && eps.acceptsString("ba") && eps.acceptsString("abbaab"),
  "(a|b)* accepts mixed strings");

// ── epsilon elimination preserves the language ──
const noEps = eps.removeEpsilon();
that(!noEps.hasEpsilon(), "removeEpsilon() leaves no epsilon transitions");
equals(JSON.stringify(eps.language(6)), JSON.stringify(noEps.language(6)),
  "same language up to length 6");
that(noEps.accept.has("I"), "I is accepting now");

// ── NFA == DFA == regex ──
const dfa = Dfa.parse(
  "states: S A AB ABB\nalphabet: a b\nstart: S\naccept: ABB\n" +
  "S a A\nS b S\nA a A\nA b AB\nAB a A\nAB b ABB\nABB a A\nABB b S\n");
const re = parseRegex("(a|b)*abb");
for (const w of abb.language(7)) {
  if (w === "epsilon") continue;
  that(dfa.acceptsString(w), `DFA agrees on "${w}"`);
  that(matches(re, w), `regex agrees on "${w}"`);
}
for (const w of dfa.language(7))
  if (w !== "epsilon") that(abb.acceptsString(w), `NFA agrees on "${w}"`);

// ── malformed ──
that(rejects("states: S\nstart: S\nS a S"), "missing alphabet:");
that(rejects("states: S\nalphabet: a\nstart: S\nS a Q"), "transition to unknown state");

summary();

function rejects(s) { try { Nfa.parse(s); return false; } catch { return true; } }
